src/app/ipv/middleware.js

```diff
@@ -330,6 +320,50 @@ async function renderAttemptRecoveryPage(req, res) {
   });
 }
 
+async function validateSessionAndPage(req, res, pageId) {
+  // Check if the page is valid
+  if (!isValidIpvPage(pageId)) {
+    return render404(res);
+  }
+
+  // Check for clientOauthSessionId for recoverable timeout page
+  if (
+    req.session?.clientOauthSessionId &&
+    pageId === PAGES.PYI_TIMEOUT_RECOVERABLE
+  ) {
+    req.session.currentPage = PAGES.PYI_TIMEOUT_RECOVERABLE;
+    return res.render(getIpvPageTemplatePath(req.session.currentPage));
+  }
+
+  // Check if ipvSessionId is present
+  if (!req.session?.ipvSessionId) {
+    logError(
+      req,
+      {
+        pageId: pageId,
+        expectedPage: req.session?.currentPage,
+      },
+      "req.ipvSessionId is null",
+    );
+
+    return renderTechnicalError(req, res);
+  }
+
+  // Handle the unrecoverable timeout page
+  if (pageId === PAGES.PYI_TIMEOUT_UNRECOVERABLE) {
+    req.session.currentPage = PAGES.PYI_TIMEOUT_UNRECOVERABLE;
+    return res.render(getIpvPageTemplatePath(req.session.currentPage));
+  }
+
+  if (req.session.currentPage !== pageId) {
+    await handleUnexpectedPage(req, res, pageId);
```

Comment:
Hmm can we just `return await` this rather than this function returning true/false?